source/parser.cc

```diff
@@ -0,0 +1,253 @@
+	AstNode* ref = sv[0].get<ParseTreeNode>().get();
+	vector<AstNode*> vars;
+	for (auto i = 1u; i < sv.size(); i+=1)
+		vars.push_back(sv[i].get<ParseTreeNode>().get());
+
+	if (ref->to_string() == "print")
+		return new PrintNode(vars);
+	else
+		return new FunCall(vars, ref);
+	
+}
+
+//parse information for smurf
+void setup_ast_generation(parser &parser) {
+
+	parser["vDec"] = [](const SemanticValues &sv) {
+		AstNode *n = assign(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["assignment"] = [](const SemanticValues &sv) {
+		AstNode *n = assign(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["boolean_expression"] = [](const SemanticValues &sv) {
+		AstNode *n = bin_op(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["arithmetic_expression"] = [](const SemanticValues &sv) {
+		AstNode *n = bin_op(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["mult_term"] = [](const SemanticValues &sv) {
+		AstNode *n = bin_op(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["ifexpression"] = [](const SemanticValues &sv) {
+		AstNode *n = ifElse(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["identifier"] = [](const SemanticValues &sv) {
+		return ParseTreeNode(new VariableValue(sv.str()));
+	};
+
+	parser["variablereference"] = [](const SemanticValues &sv) {
+		return ParseTreeNode(new VariableValue(sv.str()));
+	};
+
+	parser["number"] = [](const SemanticValues &sv) {
+	        return ParseTreeNode(new IntegerNode(atoi(sv.c_str())));
+	};
+
+	parser["add_op"] = [](const SemanticValues &sv) {
+	        return ParseTreeNode(new OpNode(sv.str()));
+	};
+
+	parser["mul_op"] = [](const SemanticValues &sv) {
+	        return ParseTreeNode(new OpNode(sv.str()));
+	};
+
+	parser["relop"] = [](const SemanticValues &sv) {
+		return ParseTreeNode(new OpNode(sv.str()));
+	};
+
+	parser["block"] = [](const SemanticValues &sv) {
+		AstNode *n = bind(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["program"] = [](const SemanticValues &sv) {
+		AstNode *n = bind(sv);
+		ParseTreeNode val = ParseTreeNode(n);
+		Interpreter* I = new Interpreter();
+		val.get()->accept(I);
+		return val;
+	};
+
+	parser["function_definition"] = [](const SemanticValues &sv) {
+		AstNode *n = funDef(sv);
+		return ParseTreeNode(n);
+	
+	};
+
+	parser["functioncall"] = [](const SemanticValues &sv) {
+		AstNode *n = funCall(sv);
+		ParseTreeNode val = ParseTreeNode(n);
+		return val;
+		
+	};
+
+}
+
+int main(int argc, const char **argv) {
+	if (argc < 2 || string("--help") == argv[1]) {
+		cout << "usage: parser [formula]" << endl;
+		return 1;
+	}
+
+	//create parser with parsing information and grammar
+	parser parser(grammar);
+	setup_ast_generation(parser);
+
+	//read in file
+	ifstream ifs;
+	ifs.open(argv[1]);
+	string toSend = "";
+	string toParse = "";
+	//remove comments and new lines
+	if (ifs.is_open()) {
+		while (getline(ifs, toParse)) {
+			int i = 0;
+			while (toParse[i] != '#' && i < toParse.size()) {
+				toSend += toParse[i];
+				i++;
+			}
+			toSend += ' ';
```

Comment:
When I first looked at this code I checked the grammar, and was surprised to see it didn't handle comments and yet tests passed. I struggled for a little trying to see how you'd pulled that off inside peglib until I came across this.

This isn't heinous but it is ugly, because it means you have lexing in two places. 

# -1